compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -472,9 +472,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         variadic_error(tcx.sess, arg.span, arg_ty, "c_uint");
                     }
                     ty::FnDef(..) => {
-                        let ptr_ty = Ty::new_fn_ptr(self.tcx, arg_ty.fn_sig(self.tcx));
-                        let ptr_ty = self.resolve_vars_if_possible(ptr_ty);
-                        variadic_error(tcx.sess, arg.span, arg_ty, &ptr_ty.to_string());
+                        let sess = tcx.sess;
+                        let span = arg.span;
+                        let (sugg_span, replace) =
+                            if let Ok(snippet) = sess.source_map().span_to_snippet(span) {
+                                (Some(span), format!("&{snippet}"))
```

Comment:
nit: you also don't need to format the suggestion as a snippet. You can use `span.shrink_to_{hi,lo}()` to get just the beginning or end of the span to add just the part you needed (in this case `&`, but in the corrected case `as ...`)

Reply:
Can you elaborate a bit more? How can I add a string to that span?
I've seen that you can do that in a method of `Diag` but I couldn't find anything.

Reply:
I'm not sure what you're asking for here. You just do the same thing you're doing here but pass `span.shrink_to_hi()` instead of `span`.

Reply:
Oh I think understand, this is so that the `^^^^^^^` are only on the suggestion?

Reply:
What happens is that instead of `f as fn()` I only get `as fn()` which would make the suggestion incorrect.

Reply:
Are you calling `.shrink_to_hi()`??

Reply:
Yes, instead of 
`                            if let Ok(fn_item) = sess.source_map().span_to_snippet(fn_item_span) {`
`                            if let Ok(fn_item) = sess.source_map().span_to_snippet(fn_item_span.shrink_to_hi()) {`
which makes sense because it's an empty span and it has nothing.

Reply:
No, that's not it. You need to stop using `span_to_snippet` at all; the whole point is that we don't need to be using a snippet at all, since rendering the expression over again is kinda redundant. 

When you're preparing the span for the suggestion below (`Some(fn_item_span)`) you use `shrink_to_hi()` there.

Reply:
Okay, is it correct now?